# Decision Tree

In [1]:
# Model Building - decision tree classification
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
df = pd.read_csv('C:\\Users\\sathv\\Downloads\\data (2).csv')

In [3]:
df.head(10)

,id,diagnosis,radius_mean,texture_mean,perimeter_mean,area_mean,smoothness_mean,compactness_mean,concavity_mean,concave points_mean,...,radius_worst,texture_worst,perimeter_worst,area_worst,smoothness_worst,compactness_worst,concavity_worst,concave points_worst,symmetry_worst,fractal_dimension_worst
0,842302,M,17.99,10.38,122.80,1001.0,0.11840,0.27760,0.30010,0.14710,...,25.38,17.33,184.60,2019.0,0.1622,0.6656,0.7119,0.2654,0.4601,0.11890
1,842517,M,20.57,17.77,132.90,1326.0,0.08474,0.07864,0.08690,0.07017,...,24.99,23.41,158.80,1956.0,0.1238,0.1866,0.2416,0.1860,0.2750,0.08902
2,84300903,M,19.69,21.25,130.00,1203.0,0.10960,0.15990,0.19740,0.12790,...,23.57,25.53,152.50,1709.0,0.1444,0.4245,0.4504,0.2430,0.3613,0.08758
3,84348301,M,11.42,20.38,77.58,386.1,0.14250,0.28390,0.24140,0.10520,...,14.91,26.50,98.87,567.7,0.2098,0.8663,0.6869,0.2575,0.6638,0.17300
4,84358402,M,20.29,14.34,135.10,1297.0,0.10030,0.13280,0.19800,0.10430,...,22.54,16.67,152.20,1575.0,0.1374,0.2050,0.4000,0.1625,0.2364,0.07678
5,843786,M,12.45,15.70,82.57,477.1,0.12780,0.17000,0.15780,0.08089,...,15.47,23.75,103.40,741.6,0.1791,0.5249,0.5355,0.1741,0.3985,0.12440
6,844359,M,18.25,19.98,119.60,1040.0,0.09463,0.10900,0.11270,0.07400,...,22.88,27.66,153.20,1606.0,0.1442,0.2576,0.3784,0.1932,0.3063,0.08368
7,84458202,M,13.71,20.83,90.20,577.9,0.11890,0.16450,0.09366,0.05985,...,17.06,28.14,110.60,897.0,0.1654,0.3682,0.2678,0.1556,0.3196,0.11510
8,844981,M,13.00,21.82,87.50,519.8,0.12730,0.19320,0.18590,0.09353,...,15.49,30.73,106.20,739.3,0.1703,0.5401,0.5390,0.2060,0.4378,0.10720
9,84501001,M,12.46,24.04,83.97,475.9,0.11860,0.23960,0.22730,0.08543,...,15.09,40.68,97.65,711.4,0.1853,1.0580,1.1050,0.2210,0.4366,0.20750


In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 32 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   id                       569 non-null    int64  
 1   diagnosis                569 non-null    object 
 2   radius_mean              569 non-null    float64
 3   texture_mean             569 non-null    float64
 4   perimeter_mean           569 non-null    float64
 5   area_mean                569 non-null    float64
 6   smoothness_mean          569 non-null    float64
 7   compactness_mean         569 non-null    float64
 8   concavity_mean           569 non-null    float64
 9   concave points_mean      569 non-null    float64
 10  symmetry_mean            569 non-null    float64
 11  fractal_dimension_mean   569 non-null    float64
 12  radius_se                569 non-null    float64
 13  texture_se               569 non-null    float64
 14  perimeter_se             5

In [6]:
len(df)
#df.diagnosis.value_counts()

569

In [7]:
df.diagnosis.value_counts()

diagnosis
B    357
M    212
Name: count, dtype: int64

In [8]:
#splitting the data into training and testing data
x = df.drop(["id", "diagnosis"], axis=1)

In [9]:
y = df.diagnosis.values

In [10]:
x

,radius_mean,texture_mean,perimeter_mean,area_mean,smoothness_mean,compactness_mean,concavity_mean,concave points_mean,symmetry_mean,fractal_dimension_mean,...,radius_worst,texture_worst,perimeter_worst,area_worst,smoothness_worst,compactness_worst,concavity_worst,concave points_worst,symmetry_worst,fractal_dimension_worst
0,17.99,10.38,122.80,1001.0,0.11840,0.27760,0.30010,0.14710,0.2419,0.07871,...,25.380,17.33,184.60,2019.0,0.16220,0.66560,0.7119,0.2654,0.4601,0.11890
1,20.57,17.77,132.90,1326.0,0.08474,0.07864,0.08690,0.07017,0.1812,0.05667,...,24.990,23.41,158.80,1956.0,0.12380,0.18660,0.2416,0.1860,0.2750,0.08902
2,19.69,21.25,130.00,1203.0,0.10960,0.15990,0.19740,0.12790,0.2069,0.05999,...,23.570,25.53,152.50,1709.0,0.14440,0.42450,0.4504,0.2430,0.3613,0.08758
3,11.42,20.38,77.58,386.1,0.14250,0.28390,0.24140,0.10520,0.2597,0.09744,...,14.910,26.50,98.87,567.7,0.20980,0.86630,0.6869,0.2575,0.6638,0.17300
4,20.29,14.34,135.10,1297.0,0.10030,0.13280,0.19800,0.10430,0.1809,0.05883,...,22.540,16.67,152.20,1575.0,0.13740,0.20500,0.4000,0.1625,0.2364,0.07678
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
564,21.56,22.39,142.00,1479.0,0.11100,0.11590,0.24390,0.13890,0.1726,0.05623,...,25.450,26.40,166.10,2027.0,0.14100,0.21130,0.4107,0.2216,0.2060,0.07115
565,20.13,28.25,131.20,1261.0,0.09780,0.10340,0.14400,0.09791,0.1752,0.05533,...,23.690,38.25,155.00,1731.0,0.11660,0.19220,0.3215,0.1628,0.2572,0.06637
566,16.60,28.08,108.30,858.1,0.08455,0.10230,0.09251,0.05302,0.1590,0.05648,...,18.980,34.12,126.70,1124.0,0.11390,0.30940,0.3403,0.1418,0.2218,0.07820
567,20.60,29.33,140.10,1265.0,0.11780,0.27700,0.35140,0.15200,0.2397,0.07016,...,25.740,39.42,184.60,1821.0,0.16500,0.86810,0.9387,0.2650,0.4087,0.12400


In [12]:
#Train & Test Split
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=0.25)

In [13]:
len(x_train)

426

In [14]:
len(x_test)

143

In [15]:
#Feature Scaling
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
x_train_sc = sc.fit_transform(x_train)
x_test_sc = sc.transform(x_test)


In [16]:
x_train_sc

array([[-0.30003685, -1.28167122, -0.37597181, ..., -1.1994018 ,
        -1.32441807, -0.98562157],
       [-0.93055584,  0.55496142, -0.96849488, ..., -1.7351982 ,
        -0.75308819, -0.97919926],
       [-0.74917366, -1.10137351, -0.73983105, ..., -0.60793232,
         0.0331008 ,  0.40587965],
       ...,
       [-0.80387622, -0.23594452, -0.74233466, ...,  0.11248024,
         0.63047422,  2.43426028],
       [-1.27143916, -1.44994908, -1.1516763 , ..., -0.10444752,
        -0.1182772 ,  2.79283944],
       [ 1.7527487 ,  1.15354981,  2.16895648, ...,  2.06938106,
         4.10077427,  0.82547079]], shape=(426, 30))

In [17]:
x_test_sc

array([[-1.30282115, -0.78404955, -1.31482878, ..., -0.63963714,
        -0.26639976, -0.13038351],
       [ 0.17213262, -1.24801565,  0.27121359, ..., -0.05438727,
        -0.15245933,  0.42086505],
       [-1.25243722, -0.84174481, -1.26100098, ..., -0.95258958,
        -0.12967124, -0.13198909],
       ...,
       [ 0.29017498,  0.68477577,  0.2311557 , ...,  0.58577718,
        -0.39987284, -0.29361731],
       [-0.42383738,  1.15354981, -0.4327205 , ...,  0.0548351 ,
        -0.15408705,  0.62263273],
       [-0.38065115,  2.46611711, -0.4327205 , ..., -0.74946631,
        -0.83610193, -1.03700008]], shape=(143, 30))

# Model Building - decision tree classification

In [24]:
from sklearn.tree import DecisionTreeClassifier
#initiate the classifier
model_dtc = DecisionTreeClassifier()
# passing the training data to the model
model_dtc.fit(x_train_sc,y_train)


DecisionTreeClassifier()

In [25]:
y_pred_dtc = model_dtc.predict(x_test_sc)

In [27]:
y_pred_dtc

array(['B', 'B', 'B', 'B', 'M', 'B', 'B', 'B', 'M', 'B', 'B', 'B', 'M',
       'B', 'B', 'B', 'M', 'B', 'M', 'M', 'B', 'B', 'B', 'B', 'B', 'B',
       'B', 'M', 'B', 'B', 'M', 'M', 'M', 'M', 'M', 'B', 'M', 'B', 'B',
       'B', 'M', 'M', 'B', 'B', 'B', 'B', 'M', 'M', 'B', 'B', 'B', 'M',
       'M', 'M', 'B', 'B', 'B', 'B', 'M', 'B', 'B', 'B', 'B', 'M', 'M',
       'B', 'B', 'B', 'B', 'B', 'B', 'M', 'M', 'B', 'B', 'M', 'B', 'M',
       'B', 'M', 'B', 'B', 'M', 'B', 'M', 'B', 'M', 'B', 'M', 'B', 'M',
       'B', 'M', 'M', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'M', 'M', 'B',
       'B', 'M', 'B', 'B', 'B', 'M', 'B', 'M', 'M', 'M', 'M', 'B', 'M',
       'B', 'M', 'B', 'B', 'M', 'B', 'M', 'B', 'M', 'M', 'B', 'B', 'B',
       'B', 'B', 'B', 'B', 'B', 'M', 'M', 'M', 'M', 'M', 'B', 'B', 'B'],
      dtype=object)

In [28]:
y_test

array(['B', 'B', 'B', 'B', 'M', 'B', 'B', 'B', 'M', 'B', 'B', 'B', 'M',
       'B', 'B', 'B', 'M', 'B', 'M', 'M', 'B', 'B', 'B', 'B', 'B', 'M',
       'B', 'M', 'M', 'B', 'M', 'M', 'M', 'M', 'M', 'B', 'M', 'B', 'B',
       'B', 'M', 'M', 'B', 'M', 'B', 'B', 'B', 'M', 'B', 'B', 'B', 'M',
       'M', 'M', 'B', 'B', 'B', 'B', 'M', 'B', 'B', 'B', 'B', 'M', 'M',
       'B', 'B', 'B', 'B', 'B', 'B', 'M', 'M', 'B', 'B', 'M', 'B', 'M',
       'B', 'M', 'B', 'B', 'M', 'B', 'M', 'M', 'M', 'B', 'M', 'B', 'M',
       'B', 'B', 'M', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'M', 'M', 'B',
       'B', 'M', 'B', 'B', 'B', 'M', 'B', 'M', 'M', 'M', 'B', 'B', 'M',
       'B', 'M', 'B', 'B', 'M', 'B', 'M', 'B', 'M', 'B', 'B', 'B', 'B',
       'B', 'B', 'B', 'B', 'B', 'M', 'M', 'B', 'M', 'M', 'M', 'B', 'B'],
      dtype=object)

almost same in comparsion let us compare them through classification metrics - accuracy_score

In [29]:
from sklearn.metrics import accuracy_score
print(round(accuracy_score(y_test,y_pred_dtc)*100,2))

93.01


accuracy_score has gone down when compared with the KNN that is nearly 98% and now this model accuracy is 93.01%

In [30]:
new_patient = [[17.99,10.38,122.8,1001,0.1184,0.2776,0.3001,0.1471,0.2419,0.07871,1.095,0.9053,8.589,153.4,0.006399,0.04904,0.05373,0.01587,0.03003,0.006193,25.38,17.33,184.6,2019,0.1622,0.6656,0.7119,0.2654,0.4601,0.1189]]

In [34]:
# feature Scaling
new_data_sc = sc.transform(new_patient)
new_data_sc

c:\ProgramData\anaconda3\Lib\site-packages\sklearn\utils\validation.py:2739: UserWarning: X does not have valid feature names, but StandardScaler was fitted with feature names
  warnings.warn(


array([[ 1.12222971, -2.11344464,  1.29686267,  1.03145384,  1.5552576 ,
         3.22135904,  2.67834919,  2.53427705,  2.22542651,  2.27223953,
         2.81871303, -0.55212826,  3.27258958,  3.17079788, -0.20247006,
         1.35484128,  0.70440189,  0.68182244,  1.15643033,  0.9416556 ,
         1.88461873, -1.37499855,  2.30680644,  2.00892905,  1.26517061,
         2.51219302,  2.04205595,  2.29085975,  2.73511679,  1.85625208]])

so here this warning is because we have given a new_data only the column values without column name of it while training we train the model with the column names too.
if the order of the new_data matches its fine but if the order does not match then the model predicts the things wrongly or will throw an error.
so in order to escape from this error 
-- new_patient_df = pd.DataFrame(new_patient, columns=X_train.columns)
-- new_patient_sc = sc.transform(new_patient_df)  try like this.

In [42]:
#predicting the class for the new patient data.
y_pred_new = model_dtc.predict(new_data_sc)
y_pred_new

array(['M'], dtype=object)

other way to not get the warning msg

In [39]:
new_patient_df = pd.DataFrame(new_patient,columns=x_train.columns)

In [40]:
new_data_sc1 = sc.transform(new_patient_df)

In [41]:
new_data_sc1

array([[ 1.12222971, -2.11344464,  1.29686267,  1.03145384,  1.5552576 ,
         3.22135904,  2.67834919,  2.53427705,  2.22542651,  2.27223953,
         2.81871303, -0.55212826,  3.27258958,  3.17079788, -0.20247006,
         1.35484128,  0.70440189,  0.68182244,  1.15643033,  0.9416556 ,
         1.88461873, -1.37499855,  2.30680644,  2.00892905,  1.26517061,
         2.51219302,  2.04205595,  2.29085975,  2.73511679,  1.85625208]])

In [44]:
#predicting the class for the new patient data.
y_pred_new1 = model_dtc.predict(new_data_sc1)
y_pred_new1

array(['M'], dtype=object)